rfc-113-expanding-draft-access-for-unauthenticated-users.md

```diff
@@ -0,0 +1,212 @@
+  from users who do not meet the criteria.
+
+These were then used to form conclusions on the level of access a
+step-by-step editor should be able to grant to a fact checker.
+
+The editor should be able to grant access to a fact checker to view the
+draft step-by-step navigation page via an auth-bypass token, this should allow
+the fact checker to bypass access limits on the content and any assets created
+for it.
+
+Content that is linked as part of a step-by-step-navigation (and assets that
+are associated with that content) may be published by any publishing application
+and by any organisation. A step-by-step editor can grant a fact checker access
+to live and draft resources, however this cannot be used to provide access to
+access limited content and assets as this would break the expectations of
+privacy.
+
+To provide access across a selection of pages an auth-bypass token would be
+stored in a session cookie this would allow maintaining this token across
+multiple requests without appending a query string (this cookie
+technique is already used for providing an [auth-bypass token to Asset
+Manager][asset-manager-auth-bypass]). To provide the auth-bypass id to
+supporting pages the Publishing API's link expansion system would be used
+for particular document and link types. As there isn't a linking system between
+assets and content it is impossible to maintain accurate links, thus an
+auth-bypass token can grant access to all draft assets for the duration of a
+session. Access limited assets must not be accessible in this circumstance
+unless the asset contains an auth-bypass id that matches the token.
+
+If there is a need to preview access limited content as part of a fact check
+then individual links will have to be provided for that content.
+
+### Consequences
+
+To support this functionality a number of changes are required in the GOV.UK
+stack, these are intended to be non-specific to the step-by-step problem and to
+support similar functionality for future GOV.UK content.
+
+- Authenticating Proxy will be changed to recognise auth-bypass tokens that are
+  present in cookies.
+- Asset Manager will be changed to recognise the presence of a flag in an
+  auth-bypass token that will grant an unauthenticated user access to a draft
+  asset that is not access limited. Fact checkers are not expected to be able
+  to exploit this access given sensitive assets should be access limited and
+  the URLs of assets are relatively unpredictable.
+- Currently Authenticating Proxy and Asset Manager do not check whether an
+  authenticated user has access to a resource if an auth-bypass token is
+  present. This introduces potential confusion when this token is provided by
+  a session cookie and could prevent an authenticated user viewing other parts
+  of draft GOV.UK. To counter this both apps will be changed to consider both
+  authentication and auth-bypass token when considering access to resources.
```

Comment:
In cases where both are supplied, which will take precedent?

Reply:
I don't believe there's a case for precedence so both can apply independently.

This is how I anticipate the logic to be in content store which takes both into account:

```
if we have a token and it's contains the root auth_bypass_id of the content
  show content
else if we're authenticated and one of the users / organisations in the access limit
  show content
else if the content doesn't have access limit and has links which include the auth_bypass_id that matches the users auth_bypass_id
  show content
else if user doesn't have a auth_bypass_id
  show content (we're assuming this is the live stack)
else
  forbidden
end
```